### PR TITLE
[66508] Add support consecutive availability

### PR DESCRIPTION
**Pull request by mrashed-dev**
# Description
This PR enables support for the `/calendars/availability/consecutive` endpoint. 

# Usage
To check for consecutive availability:
```javascript
const params = {
  startTime: 1590454800,
  endTime: 1590780800,
  interval: 5,
  duration: 30,
  buffer: 30,
  emails: [['jane@email.com'], ['swag@nylas.com']],
  open_hours: [
    {
      emails: ['swag@nylas.com', 'jane@email.com'],
      days: [0],
      timezone: 'America/Chicago',
      start: '10:00',
      end: '14:00',
      object_type: 'open_hours',
    },
  ],
};

const consecutiveAvailability = await nylas.calendars.consecutiveAvailability(params);
```

# License
I confirm that this contribution is made under the terms of the MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Review by philrenaud — APPROVED** [review]
cc @tbadun  / @jseller  -- could you give the logic a review and verify that this looks as expected?

File: __tests__/calendar-restful-model-collection-spec.js [hunk 1]
```diff
@@ -166,6 +166,60 @@ describe('CalendarRestfulModelCollection', () => {
     });
   });
 
+  test('[CONSECUTIVE AVAILABILITY] should fetch results with params', done => {
+    const params = {
+      startTime: '1590454800',
+      endTime: '1590780800',
+      interval: 5,
+      duration: 30,
+      emails: [{ name: 'Jane', email: 'jane@email.com' }],
```
Comment (tbadun) [inline __tests__/calendar-restful-model-collection-spec.js:175]: currently the API just accepts lists of lists of emails, e.g.,
```
[["jane@email.com" ],["bill@email.com","ted@email.com"]]
```
looks for consecutive slots where jane is free and biill and ted are both free

File: __tests__/calendar-restful-model-collection-spec.js [hunk 2]
```diff
@@ -166,6 +166,60 @@ describe('CalendarRestfulModelCollection', () => {
...
+      open_hours: [
+        {
+          emails: [{ name: 'Nylas', email: 'swag@nylas.com' }],
```
Comment (tbadun) [inline __tests__/calendar-restful-model-collection-spec.js:178]: similar here - plus you'll get an error of the open_hours contains an email that does not exist in the email or free_busy lists

Comment (tbadun) [reply __tests__/calendar-restful-model-collection-spec.js:178]: unless there is some other parsing step idk

File: __tests__/calendar-restful-model-collection-spec.js (hunk 1 above)
Comment (mrashed-dev) [reply __tests__/calendar-restful-model-collection-spec.js:175]: @tbadun Ah gotcha, I think the documentation is wrong since that's where I got this from: https://developer.nylas.com/docs/api/#post/calendars/availability/consecutive

So just curious, will the first array (jane) always only have one email where as the second array has all the other participants we want to check against?

File: __tests__/calendar-restful-model-collection-spec.js (hunk 2 above)
Comment (mrashed-dev) [reply __tests__/calendar-restful-model-collection-spec.js:178]: @tbadun thank you for letting me know about that error, I'll look into maybe having a check to ensure that open_hours emails all match the emails as found in the emails element in the main object. Is the email in the `open_hours` also a 2D array for consecutive?

File: __tests__/calendar-restful-model-collection-spec.js (hunk 1 above)
Comment (tbadun) [reply __tests__/calendar-restful-model-collection-spec.js:175]: yeah, the docs are a bit off unfortunately, I flagged it with doc team a while ago but should follow up.

within the array of arrays, you can have as many arrays as you want, each sub-array will contain as many emails as are required to check for a given meeting. 

```
"emails": [["jane@email.com" ],["bill@email.com"],["ted@email.com"]]
```
will look for 3 consecutive timeslots: one with jane, one with bill one with ted (arbitrary order)

```
"emails": [["jane@email.com" ,"bill@email.com","ted@email.com"]]
```
will look for 1 timeslot where jane, bill, and ted are all available

basically, any sub-array will check for the availability of just those individuals.

File: __tests__/calendar-restful-model-collection-spec.js [hunk 3]
```diff
@@ -166,6 +166,60 @@ describe('CalendarRestfulModelCollection', () => {
...
+          days: ['0'],
```
Comment (tbadun) [inline __tests__/calendar-restful-model-collection-spec.js:179]: this is also wrong in the docs, this should be a list of integer(s)

File: __tests__/calendar-restful-model-collection-spec.js [hunk 4]
```diff
@@ -166,6 +166,60 @@ describe('CalendarRestfulModelCollection', () => {
     });
   });
 
+  test('[CONSECUTIVE AVAILABILITY] should fetch results with params', done => {
+    const params = {
+      startTime: '1590454800',
+      endTime: '1590780800',
```
Comment (tbadun) [inline __tests__/calendar-restful-model-collection-spec.js:172]: also both integers

File: src/models/calendar-restful-model-collection.ts [hunk 5]
```diff
@@ -105,4 +105,68 @@ export default class CalendarRestfulModelCollection extends RestfulModelCollecti
...
+        path: `/calendars/availability/consecutive`,
+        body: {
+          emails: options.emails,
+          duration_minutes: options.duration,
+          interval_minutes: options.interval,
+          start_time: options.startTime || options.start_time,
+          end_time: options.endTime || options.end_time,
+          free_busy: options.free_busy || [],
+          open_hours: options.open_hours,
+        },
```
Comment (tbadun) [inline src/models/calendar-restful-model-collection.ts:157]: You can also add `buffer` as a potential input to the call although it's not in the docs yet. It takes a non-negative integer or float representing the buffer minutes to add to busy time. 

If, for example, you send `buffer: 30`, there will be a 30 minute buffer before and after each individual's existing meetings/busy time that will be counted as busy.

File: src/models/calendar-restful-model-collection.ts [hunk 6]
```diff
@@ -105,4 +105,68 @@ export default class CalendarRestfulModelCollection extends RestfulModelCollecti
...
+        }>;
+        days: string[];
+        timezone: string;
+        start: string;
+        end: string;
+        object_type: string;
+      }>;
+    },
+    callback?: (error: Error | null, data?: { [key: string]: any }) => void
+  ) {
```
Comment (tbadun) [inline src/models/calendar-restful-model-collection.ts:144]: in case i missed any in the notes above:
```suggestion
  consecutiveAvailability(
    options: {
      emails: Array<string[]>;
      duration: number;
      interval: number;
      start_time?: number;
      startTime?: number;
      end_time?: number;
      endTime?: number;
      free_busy?: Array<{
        email: string;
        object: string;
        time_slots: Array<{
          object: string;
          status: string;
          start_time: number;
          end_time: number;
        }>;
      }>;
      open_hours: Array<{
        emails: Array<string[]>;
        days: number[];
        timezone: string;
        start: number;
        end: number;
        object_type: string;
      }>;
    },
    callback?: (error: Error | null, data?: { [key: string]: any }) => void
  ) {
```

File: __tests__/calendar-restful-model-collection-spec.js (hunk 2 above)
Comment (tbadun) [reply __tests__/calendar-restful-model-collection-spec.js:178]: in open_hours emails will just be a list/1D array

File: __tests__/calendar-restful-model-collection-spec.js (hunk 1 above)
Comment (mrashed-dev) [reply __tests__/calendar-restful-model-collection-spec.js:175]: Ahhhh perfect explanation, thanks!

File: __tests__/calendar-restful-model-collection-spec.js (hunk 2 above)
Comment (mrashed-dev) [reply __tests__/calendar-restful-model-collection-spec.js:178]: Fantastic, thanks!

File: src/models/calendar-restful-model-collection.ts (hunk 5 above)
Comment (mrashed-dev) [reply src/models/calendar-restful-model-collection.ts:169]: Added in, thank you!

File: src/models/calendar-restful-model-collection.ts (hunk 6 above)
Comment (mrashed-dev) [reply src/models/calendar-restful-model-collection.ts]: Yup, got it in, thank you!

File: src/models/calendar-restful-model-collection.ts [hunk 7]
```diff
@@ -105,4 +105,80 @@ export default class CalendarRestfulModelCollection extends RestfulModelCollecti
...
+    }
+
+    return this.connection
+      .request({
+        method: 'POST',
+        path: `/calendars/availability/consecutive`,
+        body: {
+          emails: options.emails,
+          duration_minutes: options.duration,
+          interval_minutes: options.interval,
```
Comment (tbadun) [inline src/models/calendar-restful-model-collection.ts:163]: does it make sense to add other naming options?, e.g., 
```suggestion
          duration_minutes: options.duration | options.duration_minutes | options.durationMinutes,
          interval_minutes: options.interval | options.interval_minutes | options.intervalMinutes,
```

**Review by tbadun — APPROVED** [review]
lgtm - love the new test for open_hours, maybe an expansion of duration/interval input names if it makes sense but not necessary

File: src/models/calendar-restful-model-collection.ts (hunk 7 above)
Comment (mrashed-dev) [reply src/models/calendar-restful-model-collection.ts:163]: Technically, yes, but this is the "old" way of handling camelCase vs. snake_case. Our new refactor (latest PR is #274) will be opting for camelCase always while searlizing/deserializing appropriately. For the sake of consistency of the current way the SDK does things I've opted to keep it the same for now.

Comment (tbadun) [reply src/models/calendar-restful-model-collection.ts:163]: perfect - looks like you're gtg then 😄
